test/solvers/functional/default_functionals_test.py

```diff
@@ -0,0 +1,319 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Test for the default functionals."""
+
+# External
+import numpy as np
+import pytest
+
+# Internal
+import odl
+from odl.util.testutils import all_almost_equal, almost_equal, noise_element
+
+# Places for the accepted error when comparing results
+PLACES = 8
+
+
+def test_L1_norm():
+    """Test the L1-norm."""
+
+    n = 10
+    space = odl.rn(n)
+    func = odl.solvers.L1Norm(space)
+    x = noise_element(space)
+
+    # Evaluation of the functional
+    expected_result = np.sum(np.abs(x))
+    assert almost_equal(func(x), expected_result, places=PLACES)
+
+    # The gradient is the sign-function
+    expected_result = np.sign(x)
+    assert all_almost_equal(func.gradient(x), expected_result,
+                            places=PLACES)
+
+    # The proximal operator
+    sigma = np.random.rand()
+    # Explicit computation:      |  x_i - sigma, x_i > sigma
+    #                      z_i = {  0,           -sigma <= x_i <= sigma
+    #                            |  x_i + sigma,     x_i < -sigma
+    tmp = space.zero()
+    orig = x.asarray()
+    tmp[orig > sigma] = orig[orig > sigma] - sigma
+    tmp[orig < -sigma] = orig[orig < -sigma] + sigma
+    expected_result = space.element(tmp)
+    assert all_almost_equal(func.proximal(sigma)(x), expected_result,
+                            places=PLACES)
+
+    # The convex conjugate functional
+    cc_func = func.convex_conj
+
+    # Evaluation of convex conjugate
+    # Explicit calculation: 0 if |x|_1 < 1, infty else
+    norm_larger_than_one = 2.0 * space.one()
+    assert cc_func(norm_larger_than_one) == np.inf
+
+    norm_less_than_one = 0.9 * 1.0 / n * space.one()
+    assert cc_func(norm_less_than_one) == 0
+
+    # The gradient of the convex conjugate (not implemeted)
+    with pytest.raises(NotImplementedError):
+        cc_func.gradient
+
+    # The proximal of the convex conjugate
+    # Explicit computation: x / max(1, |x|)
+    expected_result = x / np.max((space.one(), np.abs(x)), axis=0)
+    assert all_almost_equal(cc_func.proximal(sigma)(x), expected_result,
+                            places=PLACES)
+
+    # The biconjugate, which is the functional itself since it is proper,
+    # convex and lower-semicontinuous
+    expected_functional = odl.solvers.L1Norm(space)
+    cc_cc_func = cc_func.convex_conj
+    assert isinstance(cc_cc_func, type(expected_functional))
+
+
+def test_L2_norm():
+    """Test the L2-norm."""
+
+    n = 10
+    space = odl.rn(n)
+    func = odl.solvers.L2Norm(space)
+    x = noise_element(space)
+
+    # Evaluation of the functional
+    expected_result = np.sqrt(np.sum(x**2))
+    assert almost_equal(func(x), expected_result, places=PLACES)
+
+    # The gradient
+    if x.norm() == 0:
```

Comment:
Also this should be explicitly tested:

```
with pytest.raises(ValueError):
    func.gradient(func.domain.zero())
```

The `if x.norm() == 0` statement is not really needed, the chance that a 10 element vector is all zeros is somewhere in the vicinity of 1 : 10^500